pkg/sql/distsqlrun/mergejoiner.go

```diff
@@ -84,90 +87,164 @@ func newMergeJoiner(
+// valuesProcessor ran out of rows or encountered an error. It is ok for err to
+// be nil indicating that we're done producing rows even though no error
+// occurred.
+func (m *mergeJoiner) producerMeta(err error) ProducerMetadata {
+	var meta ProducerMetadata
+	if !m.closed {
+		meta = ProducerMetadata{Err: err}
+		if err == nil {
+			meta.TraceData = getTraceData(m.ctx)
 		}
+		// We need to close as soon as we send producer metadata as we're done
+		// sending rows. The consumer is allowed to not call ConsumerDone().
+		m.close()
 	}
-	log.VEventf(ctx, 2, "exiting merge joiner run")
+	return meta
 }
 
-// outputBatch outputs all the rows corresponding to a streamMerger batch (the
-// cross-product of two groups of matching rows).
-//
-// Returns true if more batches are available and needed. If false is returned,
-// the caller should drain the inputs (as the termination condition might have
-// been dictated by the consumer saying that no more rows are needed) and close
-// the output.
-func (m *mergeJoiner) outputBatch(
-	ctx context.Context, cancelChecker *sqlbase.CancelChecker,
-) (bool, error) {
-	leftRows, rightRows, err := m.streamMerger.NextBatch(m.evalCtx)
-	if err != nil {
-		return false, err
-	}
-	if leftRows == nil && rightRows == nil {
-		return false, nil
-	}
-	var matchedRight []bool
-	if m.joinType == fullOuter || m.joinType == rightOuter {
-		matchedRight = make([]bool, len(rightRows))
+func (m *mergeJoiner) Next() (sqlbase.EncDatumRow, ProducerMetadata) {
+	if m.maybeStart("merge joiner", "MergeJoiner") {
+		m.evalCtx = m.flowCtx.NewEvalCtx()
+		m.cancelChecker = sqlbase.NewCancelChecker(m.ctx)
+		log.VEventf(m.ctx, 2, "starting merge joiner run")
 	}
 
-	for _, lrow := range leftRows {
-		matched := false
-		for rIdx, rrow := range rightRows {
-			if err := cancelChecker.Check(); err != nil {
-				return false, err
-			}
-			renderedRow, err := m.render(lrow, rrow)
-			if err != nil {
-				return false, err
+	for {
+		row, meta := m.nextRow()
+		if m.closed || meta != nil {
+			return nil, *meta
+		}
+		if row == nil {
+			return nil, m.producerMeta(nil /* err */)
+		}
+
+		outRow, status, err := m.out.ProcessRow(m.ctx, row)
+		if err != nil {
+			return nil, m.producerMeta(err)
+		}
+		switch status {
+		case NeedMoreRows:
+			if outRow == nil && err == nil {
+				continue
 			}
-			if renderedRow != nil {
-				matched = true
-				if matchedRight != nil {
-					matchedRight[rIdx] = true
+		case DrainRequested:
+			// TODO(peter): Could we be calling ConsumerDone twice on the inputs?
+			// Looks like it can happen here and in response to
+			// mergeJoiner.ConsumerDone(). This also affects distinct and
+			// noopProcessor.
+			m.leftSource.ConsumerDone()
+			m.rightSource.ConsumerDone()
+			continue
+		}
+		return outRow, ProducerMetadata{}
+	}
+}
+
+func (m *mergeJoiner) nextRow() (sqlbase.EncDatumRow, *ProducerMetadata) {
+	// The loops below form a restartable state machine the iterates over a batch
```

Comment:
Elegant! Thanks for the comment. Would be good to get one more review on the refactored merge joiner state machine. @RaduBerinde, do you mind taking a quick look at this function?